deepspeed/runtime/zero/partition_parameters.py

```diff
@@ -1043,6 +1083,34 @@ def _partition_gradient(self, param, partition_buffer=None, accumulate=False):
         param.grad.data = dest_tensor_full_buffer.data
         see_memory_usage("After partitioning gradients", force=False)
 
+    def _update_status(self, param_list, new_status):
+        if len(param_list) == 0:
+            return
+
+        for param in param_list:
+            assert param.ds_status != new_status, f"Detected redundant status update, param = {param.ds_id} status = {param.ds_status}"
+            old_status = param.ds_status
+            param.ds_status = new_status
+            self._update_param_status(new_status, old_status, param.ds_numel)
+
+    def _update_param_status(self, new_status, old_status, numel):
+        if old_status == ZeroParamStatus.AVAILABLE:
+            self.avaliable_parameter_numel -= numel
+            assert self.avaliable_parameter_numel >= 0, f'available_parameter numel is negative: {self.avaliable_parameter_numel}'
+
+        if new_status == ZeroParamStatus.AVAILABLE:
+            self.avaliable_parameter_numel += numel
+
+    def _get_available_parameter_numel(self):
+        return self.avaliable_parameter_numel
+
+    def _synchronize_communication(self, param_list, handle_list):
+        for param, handle in zip(param_list, handle_list):
+            if handle is not None:
+                handle.wait()
```

Comment:
`handle.wait()` ensures the communication task is enqueued, but not neccessarily completed. 
Based on the section "Synchronous and asynchronous collective operations" of [torch.distributed](https://pytorch.org/docs/stable/distributed.html) 

So we probably need other sync mechanism.

Reply:
there's been a lot of discussion about this here https://github.com/microsoft/DeepSpeed/issues/1069#issuecomment-870185380

the `handle.wait()` will make sure no more work happens on current stream until the collective has completed so for this to be safe it either needs to be consumed in the same stream or synchronized.

@zarzen calls out below that the call is made from `self.comm_stream`, i think if we get rid of that (do the all gather from default stream so that kernel launches on default stream dont progress until the allgather completes) we won't need the additional synchronize. Like he mentions below it seems to independently decide to use its own stream so the actual allgather work isn't being done in `comm_stream` anyways

Reply:
@jfc4050 and @zarzen, thanks for your deep investigation of this issue. I am closely following your updates on #1069. 

Thanks for your explanation, I now see the value of eliminating `self.comm_stream` and doing the `handle.wait()` in default stream. I will try this out.

Reply:
np! thinking about this a little more, there could be a small amount of value to doing allgather in the comm_stream, but not for parallelizing communication - there's some memcpy operations that get done as part of allgather that could potentially be overlapped with forward pass computation and parameter gathering communication.

Doing this means that the streams would have to be carefully synchronized where appropriate though, so might not be worth the extra complexity, hopefully we can find lower hanging fruit

Reply:
That makes sense, the low benefits does not justify the complexity.